pkg/controller/dataimportcron-controller.go

```diff
@@ -0,0 +1,532 @@
+	cdiv1 "kubevirt.io/containerized-data-importer/pkg/apis/core/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// DataImportCronReconciler members
+type DataImportCronReconciler struct {
+	client         client.Client
+	uncachedClient client.Client
+	recorder       record.EventRecorder
+	scheme         *runtime.Scheme
+	log            logr.Logger
+}
+
+const (
+	dataImportControllerName      = "dataimportcron-controller"
+	annCronJobActiveJobName       = "cronJobActiveJobName"
+	annCronJobDigest              = "cronJobDigest"
+	cronJobImageStreamTermMessage = "imagestream"
+	labelDataImportCronName       = "dataimportcron-name"
+	recentImportsToKeepPerCronJob = 3
+)
+
+// Reconcile loop for DataImportCronReconciler
+func (r *DataImportCronReconciler) Reconcile(_ context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithName("Reconcile")
+	dataImportCron := &cdiv1.DataImportCron{}
+	if err := r.client.Get(context.TODO(), req.NamespacedName, dataImportCron); err != nil {
+		if k8serrors.IsNotFound(err) {
+			log.Info("DataImportCron not found", "name", req.NamespacedName)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	return r.reconcileDataImportCron(dataImportCron)
+}
+
+// Reconcile DataImportCron
+func (r *DataImportCronReconciler) reconcileDataImportCron(dataImportCron *cdiv1.DataImportCron) (reconcile.Result, error) {
+	cronJob, err := r.getCronJob(dataImportCron)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	digest, err := r.getDigest(dataImportCron, cronJob)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if err := r.updateStatus(dataImportCron); err != nil {
+		return reconcile.Result{}, err
+	}
+	if r.shouldImport(dataImportCron, digest) {
+		if err := r.createImportDataVolume(dataImportCron, cronJob); err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *DataImportCronReconciler) getCronJob(dataImportCron *cdiv1.DataImportCron) (*v1beta1.CronJob, error) {
+	log := r.log.WithName("getCronJob")
+	// If no k8s cron job found, create a new one
+	cronJob := &v1beta1.CronJob{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: dataImportCron.Namespace, Name: dataImportCron.Name}, cronJob); err != nil {
+		if k8serrors.IsNotFound(err) {
+			cronJob := newCronJob(dataImportCron)
+			if err := r.client.Create(context.TODO(), cronJob); err != nil {
+				log.Error(err, "Unable to create CronJob")
+				return nil, err
+			}
+			return cronJob, nil
+		}
+		return nil, err
+	}
+	return cronJob, nil
+}
+
+func (r *DataImportCronReconciler) getDigest(dataImportCron *cdiv1.DataImportCron, cronJob *v1beta1.CronJob) (string, error) {
```

Comment:
So let me see if I understand this code correctly:
1. You check if the job is active.
2. If it is active you put the jobname on the on the cron object
3. You check if the digest is set on the job
4. If it is set, then you return a blank string for the result of getDigest?
5. You look up all the pods owned by the job
6. If there is one pod, then you look up the termination mesage
7. The termination message is either a node or skopeo digest
8. You set the found digest on the cron job
9. You return the digest

How are you reliably going to find the pods. If the digest doesn't take long to obtain the pods might be long gone before you try to read them. Or the job might no longer be active (since it finished) and you won't even try to find the pods.
Wouldn't it make much more sense to have a controller that watches for pods that have an ownerRef of a cron job, and in the reconcile loop of that controller look up the pod (we know it will exist since that triggered the reconcile) and look up the owning cron job. And do the digest update in that controller.

And in this controller, just check for the annotation, and if it is there do the logic for importing if it is different from the current digest. You are doing too many things at once in this controller IMO.